src/test/regress/sql/failure_multi_dml.sql

```diff
@@ -0,0 +1,148 @@
+SELECT citus.mitmproxy('conn.allow()');
+
+SET citus.shard_count = 2;
+SET citus.shard_replication_factor = 1; -- one shard per worker
+SET citus.next_shard_id TO 103400;
+ALTER SEQUENCE pg_catalog.pg_dist_placement_placementid_seq RESTART 100;
+
+CREATE TABLE dml_test (id integer, name text);
+SELECT create_distributed_table('dml_test', 'id');
+
+COPY dml_test FROM STDIN WITH CSV;
+1,Alpha
+2,Beta
+3,Gamma
+4,Delta
+\.
+
+SELECT citus.clear_network_traffic();
+
+---- test multiple statements spanning multiple shards, at each possible point
+
+-- fail at DELETE
+SELECT citus.mitmproxy('conn.onQuery(query="^DELETE").kill()');
+
+BEGIN;
+DELETE FROM dml_test WHERE id = 1;
+DELETE FROM dml_test WHERE id = 2;
+INSERT INTO dml_test VALUES (5, 'Epsilon');
+UPDATE dml_test SET name = 'alpha' WHERE id = 1;
+UPDATE dml_test SET name = 'gamma' WHERE id = 3;
+COMMIT;
+
+--- shouldn't see any changes after failed DELETE
+SELECT * FROM dml_test ORDER BY id ASC;
+
+-- fail at INSERT
+SELECT citus.mitmproxy('conn.onQuery(query="^INSERT").kill()');
+
+BEGIN;
+DELETE FROM dml_test WHERE id = 1;
+DELETE FROM dml_test WHERE id = 2;
+INSERT INTO dml_test VALUES (5, 'Epsilon');
+UPDATE dml_test SET name = 'alpha' WHERE id = 1;
+UPDATE dml_test SET name = 'gamma' WHERE id = 3;
+COMMIT;
+
+--- shouldn't see any changes before failed INSERT
+SELECT * FROM dml_test ORDER BY id ASC;
+
+-- fail at UPDATE
+SELECT citus.mitmproxy('conn.onQuery(query="^UPDATE").kill()');
+
+BEGIN;
+DELETE FROM dml_test WHERE id = 1;
+DELETE FROM dml_test WHERE id = 2;
+INSERT INTO dml_test VALUES (5, 'Epsilon');
+UPDATE dml_test SET name = 'alpha' WHERE id = 1;
+UPDATE dml_test SET name = 'gamma' WHERE id = 3;
+COMMIT;
+
+--- shouldn't see any changes after failed UPDATE
+SELECT * FROM dml_test ORDER BY id ASC;
+
+-- FIXME: Can't test this because the prepared transaction ID has PID in it
+
+-- fail at COMMIT (by failing PREPARE)
+-- SELECT citus.mitmproxy('conn.onQuery(query="^PREPARE").kill()');
```

Comment:
This makes me assume the query is a 2pc, can we also check to fail at commit, I would expect to see all changes after recovery.